lucetc/src/decls.rs

```diff
@@ -155,32 +168,33 @@ impl<'a> ModuleDecls<'a> {
                 }
             };
 
+            let custom_info = custom_name_for(ix, func_index, decls);
             let import_info = import_name_for(func_index, decls, bindings)?;
             let export_info = export_name_for(func_index, decls);
 
+            // We don't match on `custom_info` because export and import info are needed to
+            // determine the linkage type, and because we use different values as a fallback for
+            // exported functions and local functions.
             match (import_info, export_info) {
                 (Some(import_sym), _) => {
                     // if a function is only an import, declare the corresponding artifact import.
                     // if a function is an export and import, it will not have a real function body
                     // in this program, and we must not declare it with Linkage::Export (there will
                     // never be a define to satisfy the symbol!)
-
                     decls.declare_function(clif_module, import_sym, Linkage::Import, func_index)?;
                 }
                 (None, Some(export_sym)) => {
                     // This is a function that is only exported, so there will be a body in this
                     // artifact. We can declare the export.
+                    let export_sym = custom_info.unwrap_or(export_sym);
```

Comment:
I'm not sure we want to do this - if a function is declared an export then there's an explicit name someone's expecting to see it by, and at the least we're appending its index (an implementation detail from Lucet) to that name. We ought to be able to just use `export_sym` because duplicate exports would be rejected by wasm validation.

From there, the big issue is "what if a custom name collides with an export name?", which I think is only possible if there's an export `foo_N` for some function, and a custom function named `foo` at index N, where we'll have the export as `guest_func_foo_N` and the custom name will _also_ be `guest_func_foo_N`. I'm not sure if this is a novel error mode, because I _think_ symbols can be just numbers, and if so it would Technically Be Possible that we have this circumstance with an export `1234` with an anonymous function at index `1234`.

It probably makes sense to disambiguate `guest_func` from exported functions generally, so that we might have `guest_func_foo_1235` for an internal function, and `guest_export_foo` for an actual honest-to-goodness exported `foo`.  (To be clear, changing `guest_func` to `guest_export` would probably be best in a different PR if we can put together a test demonstrating the collision.)

Reply:
Ah yeah, good point! Undid this, the name collision stuff seems like it can be investigated outside of this PR :slightly_smiling_face: